### DISCUSSION TITLE
Add duration to device classes and SensorEntity conversions and sensor value scaling

### DISCUSSION THREAD

**Discussion by khcnz** [home-assistant/architecture#630] [Entity Models] [answered]
Love the new consolidation of conversions into the base SensorEntity! I've got an idea around how time sensors are treated for consideration.

## Context
There are many situations where a time period / time delta is a relevant sensor type and there are multiple core classes that utilize a sensor in seconds (and other time period) like this [native_unit_of_measurement=TIME_SECONDS](https://github.com/home-assistant/core/search?p=2&q=TIME_SECONDS)

In particular I've got a sensor with a device that reports the cumulative operating time of a device in seconds which I'd like to add to long term statistics and also to display the number in a more relevant way.

In many of these cases (but probably not all), it would make sense to display in lovelace by default something with a time period that is more relevant to the size of the time period. Rather than 3,264,240 seconds 37.78 days is a more relevant number to display.

These are the existing time constants
``` 
TIME_MICROSECONDS: Final = "μs"
TIME_MILLISECONDS: Final = "ms"
TIME_SECONDS: Final = "s"
TIME_MINUTES: Final = "min"
TIME_HOURS: Final = "h"
TIME_DAYS: Final = "d"
TIME_WEEKS: Final = "w"
TIME_MONTHS: Final = "m"
TIME_YEARS: Final = "y"
``` 

## Proposal Part 1 - New device class duration
Create a new device class (duration) that accepts a unit of measurement of any of the existing HA time periods 
`DEVICE_CLASS_DURATION: Final = "duration"`

## Proposal Part 2 - Convert & scale sensor values
I started writing this discussion up relating to time, but I realized this is just a special case of _unit of scale_ conversion rather than something specific to time. The same really applies to W vs kW, bit/s vs Mbit/s and many other cases. Maybe I'm missing something and this already happens but there should be a way to consolidate logic in these conversions into one place.

**Option A:** Each integration fends for itself and does conversations to units of scale that it sees fit, or asks the user what units of scale to use in a config flow. The [integration sensor](https://github.com/home-assistant/core/blob/master/homeassistant/components/integration/sensor.py) appears to do something a little like this 

**Option B:** Implement some shared methods on the SensorEntity (or elsewhere) to help with unit of scale conversions and an integration author could optionally and manually utilize them through code.

**Option C:** In the SensorEntity class as well as units of measure conversions, also do units of scale conversions and chose a best guess as an appropriate scale to use depending on the current size of the value, return this as state.

**Option D:** Add a new property 'scaled_state' and 'scaled_unit_of_measurement' to SensorEntity or display_state and display_unit_of_measurement. Depending on the device_class this this would either be a scaled value, or the existing unit converted value. Potentially allow an integration to specify if it wants to opt-in to this scale conversion or not. Use this value in the UI.

**Option E:** Don't do anything, It's a UI concern only

### Cons
- All of these approaches end up with a question about how to decide how many decimal places to display, which i don't have an answer for.
- Option A  - Lots of code duplication in core and 3rd party integrations
- Option B - Most flexible and without code duplication, but probably won't get used a lot
- Option C - Probably isn't a good approach as everything that stores/uses state such as automations and long term stats would be limited to the level of granularity returned.
- Option D - Allows integrations to opt-in to this behaviour
- Option E: Won't easily flow through to custom ui components

Relevant screenshot from my device with large cumulative seconds and Wh
![image](https://user-images.githubusercontent.com/130233/132273931-86c83d71-ed48-4fc7-b65b-f7ab0fc8bed4.png)

**Comment by khcnz** [comment]
There is some overlap to the discussion here with bandwidth sensors and units of scale as well https://github.com/home-assistant/architecture/discussions/621

**Reply by raman325** [reply]
I would use @Bre77's suggestion from that thread here as well: Provide the duration as `native_value` and unit as `native_unit_of_measurement` and then have the sensor base class do the appropriate conversion into a more friendly state and unit.

The only problem is it would increase automation complexity if the units keep changing. One thought is to add `native_value` and `native_unit_of_measurement` attributes to any entities where the `native_value` doesn't match the `state` and `native_value` is a non complex value (e.g. not a `datetime`)

**Comment by garbled1** [comment]
I would desperately like this for the venstar integration.  The device has a daily runtime sensor, that tracks how many minutes per day the cooling/heating/whatever ran.  This is perfect for a long-running statistic, as it can show over a period of months how much you are using the heating/etc.

**Comment by elupus** [comment]
My NIBE custom integration for heat pumps also need this.

**Reply by elupus** [reply]
The duration device class that is.

**Comment by MartinHjelmare** [comment, marked as answer]
Duration device class has been implemented.
